packages/adminrouter/extra/src/README.md

```diff
@@ -535,8 +552,225 @@ DC/OS IAM relies on JSON Web Tokens for transmitting security information betwee
+generic tests and helpers and a testing framework using YAML configuration files
+was developed.
+
+#### Generic tests
+Tests shared by most of the endpoints are as follow:
+* `generic_no_slash_redirect_test` - test if request for location without
+  trailing slash is redirected to the one that ends with `/`
+* `generic_response_headers_verify_test` - test if response sent by AR is correct,
+  this mostly involves checking response headers
+* `generic_upstream_headers_verify_test` - test if the request sent by NGINX to
+  the upstream is correct. Again - this is mostly about the headers now.
+* `generic_correct_upstream_dest_test` - test if the upstream request done by
+  Admin Router is sent to the correct upstream
+* `generic_correct_upstream_request_test` - test if path component of the request
+  sent upstream by AR along with HTTP version is correct
+* `generic_location_header_during_redirect_is_adjusted_test` - test if `Location`
+  headers are being rewritten/adjusted by NGINX in a situation when the upstream
+  performs a redirect
+
+All of these tests can be found in `test-harness/modules/generic_test_code/\*.py`
+together with extensive docstrings which explain all the call parameters.
+
+#### Helpers
+Helpers shared by most of the tests are as follow:
+* `overriden_file_content` - temporarily, while in the scope of the context manager,
+  substitute the contents of the given file with a given content. At the end of
+  the context, the original file contents are restored.
+* `verify_header` and `header_is_absent` - these functions check if the given
+  list of header tuples contains or does not contain given header.
+* `GuardedSubprocess` - launches given ar instance and makes sure that it is
+  terminated at the end of the context.
+* `iam_denies_all_requests` - makes sure that within of the context, IAM denies
+  all PolicyQuery requests.
+* `ping_mesos_agent` - tests if agent with a given ID (`mocker.endpoints.mesos.AGENT1_ID`
+   by default) can be reached.
+* `repo_is_ee` - function, upon which, all remaining test harness code
+  determines repository flavour
+* `assert_iam_queried_for_uid` - asserts that within the context, IAM has been
+  queried for given UID
+* `assert_endpoint_response` - asserts response code and log messages in Admin
+  Router stderr for request against specified path
+
+Please check corresponding helpers' source code for more information,
+all of them are pretty well documented using docstrings.
+
+This list does not include LogCatcher helper which is complex enough to
+deserve a separate paragraph.
+
+#### YAML configuration
+Generic tests are useful when one wants to re-use certain tests, but with
+dozens of tests using the generic tests, tests can still become hard to read
+and maintain. On top of that, the fact that tests need to be differentiated
+by repository flavour (Open|EE) and by AR type (master|slave), makes tests even
```

Comment:
People would just keep reinventing stuff, which would eventually lead to having multiple ways of doing things, which in turn would result in some of these things being rarely used or not used at all, which would result in the code bitroting.